pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,146 @@
+package subnet
+
+import (
+	"os"
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+var (
+	log                       = logger.Log
+	MarkedForDelete           = true
+	EnforceRevisionCheckParam = false
+	ResourceTypeSubnet        = common.ResourceTypeSubnet
+	NewConverter              = common.NewConverter
+)
+
+type SubnetService struct {
+	common.Service
+	SubnetStore *SubnetStore
+}
+
+var subnetService *SubnetService
+
+// GetSubnetService get singleton SubnetService instance, subnet/subnetset controller share the same instance.
+// GetSubnetService isn't thread-safe.
+func GetSubnetService(service common.Service) *SubnetService {
+	if subnetService == nil {
+		var err error
+		if subnetService, err = InitializeSubnetService(service); err != nil {
+			log.Error(err, "failed to initialize subnet commonService")
+			os.Exit(1)
+		}
+	}
+	return subnetService
+}
+
+// InitializeSubnetService initialize Subnet service.
+func InitializeSubnetService(service common.Service) (*SubnetService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeSubnetCRUID: indexFunc}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, subnetService.SubnetStore)
```

Comment:
Perhaps this would not properly function, refer [https://github.com/vmware-tanzu/nsx-operator/pull/160/files#diff-56a01862f16c1cc9248df82cc17bcd7adcb1fe24442da3fb078cfabee9b9c7e9R158](url)

Reply:
Thanks, predicates added.

Reply:
Rebased and updated, thanks.